### DISCUSSION TITLE
Need Help: My Yourls Short URL generation is taking around 10 seconds!!

### DISCUSSION THREAD

**Discussion by Sidgut** [YOURLS/YOURLS#2932] [Q&A]
Hi Guys, Can anyone help I am using, URL shortening method as 62 & have around 1Million short URLs in the database, slowly & slowly the url generation has turned slow & now it has reached to a point that it now takes around 10secs to generate a short url. I have even tried via the dashboard & the API the result is the same.

Ever since I had started using Yourls (4yrs back) I have not faced any issues, I haven't upgraded the system I am currently using v 1.7.2 & also I have not touched my mysql database ever since.

Can anyone guide me as to what can I do to improve the performance? 

Would appreciate any references, threads, comments to solve the issue.

**Comment by ozh** [comment]
I guess you have a plugin to generate random short URLs with a fixed number of chars ?

**Reply by Sidgut** [reply]
Yes, I am using the plugin to generate short url with random short url with fixed numbers.. I had even updated the length from 5 characters to 6 characters & even ugraded my VPS but still there was hardly any significant change that I could see, the time to generate the short url was more or less the same.

Also I have seen that while generating the url the system fetches the title etc through a curl request, does that also slows down the response time or that happens after the url has been added in the db?

**Reply by ozh** [reply]
Yes, YOURLS fetches the remote page to get its title, but this should be a fast process. There's a plugin that disables this behavior if you want to make sure.

How fast is your setup if you disable the random short URL plugin ?

**Reply by Sidgut** [reply]
I just tried disabling the random short URL plugin, there is no change in the speed, later I tried disabling multiple short URLs for a same long URL by setting define( 'YOURLS_UNIQUE_URLS', false ); in the config file after that, there was a significant improvement in the link generation, I think the issue is in indexing the database.

Can you guide as to what I should do to optimize this..

**Reply by ozh** [reply]
For how long is the following query running on your setup : 
```SQL
SELECT * FROM `yourls_url` WHERE `keyword` = 'something'
```

**Reply by Sidgut** [reply]
It has been since August of 2016, that this query has been running. Would appreciate if you can guide steps to optimize the system?

Also just for your knowledge the Dashboard says : _Overall, tracking 981,654 links, 28,955,596 clicks, and counting!_

**Reply by ozh** [reply]
Yes. I'm trying to guide you to optimize. Hence my question. I rephrase : how many seconds does this query takes to run?

**Reply by Sidgut** [reply]
Oh! my bad, I didn't understand your question right, so I just checked on an average it takes around 72ms to run the Select query.

**Reply by ozh** [reply]
Add `define('YOURLS_DEBUG', true);` to your `config.php` and in the admin area shorten a new URL. In the footer you should see debug messages with queries and run time : what does it say ?

**Reply by Sidgut** [reply]
I had enabled the debug mode & here is the debug messages: 
> DB driver: pdo
> SELECT `option_name`, `option_value` FROM `yourls_options` WHERE 1=1
> SELECT COUNT(keyword) as count, SUM(clicks) as sum FROM `yourls_url` WHERE 1=1 
> SELECT * FROM `yourls_url` WHERE 1=1  ORDER BY `timestamp` desc LIMIT 0, 15;

Screenshot: https://prnt.sc/12sfee5, though I was not able to locate the run time for queries. Let me know if any other information is required from my end.

**Reply by ozh** [reply]
Screenshot using latest release:
![image](https://user-images.githubusercontent.com/223647/117815901-3d501980-b266-11eb-99ed-78ec0dc705a4.png)

I strongly suggest you backup your DB and update YOURLS. If something goes wrong you can revert to 1.7.2 using DB backup so it's risk free.

**Reply by Sidgut** [reply]
Okay, I'll update to the latest YOURLS & let you know.

**Comment by ozh** [comment]
For the record : on my totally unoptimized test box (stock XAMPP mysql.ini completely untweaked) I don't notice anything slow after inserting 4.5 millions of records (running latest release)

- Inserting a new link is not noticeably slower
- Displaying the admin page is as fast as usual
- Searching for a particular term takes a couple seconds longer, which I guess is expected (around 3.6 seconds)

```
Connected to database yourls_ozh on localhost 
SQL perform: SELECT option_name, option_value FROM yourls_options WHERE 1=1 (0.00040 s)
Check for new version: no
SQL perform: SELECT COUNT(keyword) as count, SUM(clicks) as sum FROM `yourls_url` WHERE 1=1  (1.01134 s)
SQL perform: SELECT COUNT(keyword) as count, SUM(clicks) as sum FROM `yourls_url` WHERE 1=1  AND CONCAT_WS('',`keyword`,`url`,`title`,`ip`) LIKE ('%dr%') (3.60477 s)
SQL perform: SELECT * FROM `yourls_url` WHERE 1=1  AND CONCAT_WS('',`keyword`,`url`,`title`,`ip`) LIKE ('%dr%') ORDER BY `timestamp` desc LIMIT 0, 15; (0.01439 s)
```

**Comment by ozh** [comment]
Interestingly though, the more results returned by search, the faster the query - less results make the search slower, no result at all makes the longest : 

No result:
`SQL perform: SELECT * FROM `yourls_url` WHERE 1=1  AND CONCAT_WS('',`keyword`,`url`,`title`,`ip`) LIKE ('%dddd%') ORDER BY `timestamp` desc LIMIT 0, 15; (33.04203 s)`

40 results 12 seconds, 5000 results 3 seconds. Not sure why but that something I'll investigate.

**Reply by Sidgut** [reply]
I'll be testing with the upgrade in a couple of days, will keep you posted with the update, in the meantime you can also investigate & let me know if you find any bottleneck or patch.
